### DISCUSSION TITLE
Okio in multiplatform iOS & Android

### DISCUSSION THREAD

**Discussion by yoman07** [square/okio#1111] [Q&A]
Hi,
how to use Okio in kotlin multiplatform? I want to use openZip function, but I couldn't find in documentation how to properly get FileSystem. Most of examples use FileSystem.System but it won't work. I get "Unresolved reference: SYSTEM" I am able to get only FileSystem.SYSTEM_TEMPORARY_DIRECTORY.

**Comment by JakeWharton** [comment]
Zip support is currently only available for the JVM/Android. The SYSTEM filesystem is only available for JVM/Android and native as there is no appropriate filesystem for JS (browser).

Until #980 lands and is released you will have to actual/expect the SYSTEM filesystem yourself to access it from a common source set.

**Reply by pathconnected** [reply]
Is deflate & zip still unsupported on android / iOS?

**Reply by JakeWharton** [reply]
They have always been supported on Android. Native targets were added for deflate, gzip, and zip in 3.9.

**Reply by pathconnected** [reply]
Thank you. So the documentation in https://square.github.io/okio/multiplatform/ is outdated?

```
Compression (Deflater, Inflater, Gzip)

JVM-only.
```

**Reply by JakeWharton** [reply]
Yep

**Reply by JakeWharton** [reply]
Updated

**Comment by avenwu** [comment]
+1, Same error when trying to use the `FileSystem.System` API. 

According to the okio changelog, the v3.1.0  support [Hierarchical project structure](https://kotlinlang.org/docs/multiplatform-hierarchy.html).

Any help?

---

<img width="730" alt="截屏2022-06-09 15 34 44" src="https://user-images.githubusercontent.com/1622234/172791996-8788ce5c-31e5-4e98-82a4-faff770fdb21.png">
<img width="592" alt="截屏2022-06-09 15 30 12" src="https://user-images.githubusercontent.com/1622234/172792026-1e306a1a-f01f-4b38-9d16-43edf74871d6.png">

**Reply by avenwu** [reply]
Define a custom actual/expect can avoid the error. see https://github.com/square/okio/discussions/1115

**Reply by oldergod** [reply]
Any chance you could publish a demo repo?

**Reply by avenwu** [reply]
> Any chance you could publish a demo repo?

Just create a demo project by AndroidStudio can reproduce the case, nothing else required. Are you trying to fix this reference error?

**Comment by darronschall** [comment]
Like this issue describes, I too am unable to use `FileSystem.SYSTEM` from inside the `commonMain` source set out of the box (using Okio 3.2.0). It _is_ available in both `androidMain` and `iosMain` though. A simple expect/acutal wrapper does the trick:

```kotlin
// in commonMain platformFileSystem.kt
import okio.FileSystem

expect fun platformFileSystem(): FileSystem
```

```kotlin
// in androidMain platformFileSystem.kt
import okio.FileSystem

actual fun platformFileSystem(): FileSystem { return FileSystem.SYSTEM }
```

```kotlin
// in iosMain platformFileSystem.kt
import okio.FileSystem

actual fun platformFileSystem(): FileSystem { return FileSystem.SYSTEM }
```

Then, in `commonMain` code, use `val fileSystem = platformFileSystem()`.
